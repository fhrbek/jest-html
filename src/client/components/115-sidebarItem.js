// @flow

/* eslint-env browser */
/* eslint-disable react/no-multi-comp, react/prefer-stateless-function */

import React from 'react';
import { Link } from 'react-router-dom';
import {
  Icon,
  hoverable,
  flexContainer,
  flexItem,
  darken,
  lighten,
  isHintShown,
  hintDefine,
  hintShow,
} from 'giu';
import { UI } from '../gral/constants';
import { isWaiting } from '../gral/helpers';

let fHintAlreadyShown = false;

// ==========================================
// Declarations
// ==========================================
type Props = {
  id: string,
  label: string,
  dirty?: boolean,
  deleted?: boolean,
  icon: string,
  link: string,
  fSelected: boolean,
  showBaseline?: () => any,
  hideBaseline?: () => any,
  saveAsBaseline?: (snapshotId: string) => any,
  // hoverable HOC
  hovering: any,
  onHoverStart: Function,
  onHoverStop: Function,
};

// ==========================================
// Component
// ==========================================
class SidebarItem extends React.Component<Props> {
  fDirtyIconShown: boolean;

  componentDidMount() {
    this.hintIfNeeded();
  }
  componentDidUpdate() {
    this.hintIfNeeded();
  }

  // ------------------------------------------
  render() {
    const { id, deleted, fSelected } = this.props;
    const fHovered = this.props.hovering === id;
    return (
      <Link
        id={id}
        to={this.props.link}
        style={style.outer({ fSelected, fHovered, fDeleted: deleted })}
        onMouseEnter={this.props.onHoverStart}
        onMouseLeave={this.props.onHoverStop}
      >
        <div style={flexItem('0 0 20px')}>
          <Icon icon={this.props.icon} />
        </div>
        <div style={flexItem(1)}>
          {this.props.label} {deleted && <i>(deleted)</i>}
        </div>
        {this.renderDirtyIcon()}
      </Link>
    );
  }

  renderDirtyIcon() {
    this.fDirtyIconShown = Boolean(this.props.dirty) && !this.props.deleted;
    if (!this.fDirtyIconShown) return null;
    const { saveAsBaseline, showBaseline, hideBaseline } = this.props;
    return (
      <div
        className="jh-dirty-icon"
        style={flexItem('0 0 15px', { textAlign: 'right' })}
      >
        <DirtyIcon
          id={this.props.id}
          fSelected={this.props.fSelected}
          fSnapshot={!!showBaseline}
          onHoverStart={showBaseline}
          onHoverStop={hideBaseline}
          onClick={ev => {
            saveAsBaseline && saveAsBaseline(ev.target.id);
            hideBaseline && hideBaseline();
          }}
        />
      </div>
    );
  }

  // ------------------------------------------
  hintIfNeeded() {
    if (!this.fDirtyIconShown) return;
    if (fHintAlreadyShown || isHintShown() || isWaiting()) return;
    const elements = () => {
      const out = [];
      const node = document.getElementsByClassName('jh-dirty-icon')[0];
      if (node) {
        const bcr = node.getBoundingClientRect();
        const x = bcr.left + bcr.width + 70;
        const y = bcr.top + bcr.height / 2;
        out.push({
          type: 'LABEL',
          x,
          y,
          align: 'left',
          children: (
            <span>
              Changed! <span style={style.highlight}>In snapshots,</span> hover
              to see the baseline (old) version, and click to accept as a new
              baseline
            </span>
          ),
        });
        out.push({
          type: 'ARROW',
          from: { x, y },
          to: { x: x - 70 + 5, y },
          counterclockwise: true,
        });
      }
      return out;
    };
    hintDefine('dirtyIcon', { elements });
    hintShow('dirtyIcon');
    fHintAlreadyShown = true;
  }
}

const SidebarGroup = ({ name, children }: { name: string, children?: any }) => (
  <div style={style.group}>
    <div style={style.groupTitle}>{name}</div>
    {children}
  </div>
);

class _DirtyIcon extends React.Component<{}> {
  render() {
    const {
      hovering,
      id,
      fSelected,
      fSnapshot,
      onHoverStart,
      onHoverStop,
      onClick,
    } = this.props;
    let tooltip = 'Modified since the last time jest-html was launched';
    if (hovering) {
      tooltip += '. Click to save baseline. Press ESC to dismiss this tooltip';
    } else {
      tooltip += fSnapshot
        ? '. Select this snapshot and hover to see baseline'
        : '. Click for more details on what changed';
    }
    return (
      <Icon
        id={id}
        icon="asterisk"
        onMouseEnter={fSelected ? onHoverStart : undefined}
        onMouseLeave={fSelected ? onHoverStop : undefined}
        onClick={fSnapshot ? onClick : undefined}
        style={style.dirtyIcon({ fSelected, fSnapshot, fHovering: hovering })}
        title={tooltip}
      />
    );
  }
}
const DirtyIcon = hoverable(_DirtyIcon);

// ------------------------------------------
const style = {
  outer: ({
    fSelected,
    fHovered,
    fDeleted,
  }: {
    fSelected?: boolean,
    fHovered?: boolean,
    fDeleted?: boolean,
  }) => {
    let backgroundColor;
    if (fSelected && fHovered) backgroundColor = darken(UI.color.accentBg, 10);
    else if (fHovered) backgroundColor = darken('white', 10);
    else if (fSelected) backgroundColor = UI.color.accentBg;
    let color = 'currentColor';
    if (fSelected && fDeleted) color = lighten(UI.color.accentBg, 25);
    else if (fDeleted) color = UI.color.textDim;
    else if (fSelected) color = UI.color.accentFg;
    return flexContainer('row', {
      padding: '0.3em 1em',
      cursor: 'pointer',
      color,
      backgroundColor,
      wordBreak: 'break-word',
      alignItems: 'center',
      textDecoration: 'none',
    });
  },
  dirtyIcon: ({ fSelected, fSnapshot, fHovering }) => {
    let color;
    if (fSelected && fSnapshot && fHovering)
      color = lighten(UI.color.accentBg, 25);
    else if (fSelected && fSnapshot) color = UI.color.accentFg;
    else if (fSelected) color = lighten(UI.color.accentBg, 25);
    else if (fSnapshot) color = UI.color.accentBg;
    else color = UI.color.textDim;
    return { color };
  },
  group: {
    margin: '0.5em 0',
    padding: '0.3em 0',
    backgroundColor: 'white',
    // border: '1px solid #aaa',
  },
  groupTitle: {
    color: '#999',
    padding: '0.3em 1em',
    fontSize: '0.8em',
    textTransform: 'uppercase',
  },
  highlight: {
    fontWeight: 'bold',
    color: 'yellow',
  },
};

// ==========================================
// Public
// ==========================================
export default hoverable(SidebarItem);
export { SidebarItem as _SidebarItem, SidebarGroup };
