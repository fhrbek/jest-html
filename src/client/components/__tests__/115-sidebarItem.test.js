/* eslint-env jest */
/* eslint-disable global-require, import/newline-after-import */
import React from 'react';
import renderer from 'react-test-renderer';
import { _SidebarItem as SidebarItem, SidebarGroup } from '../115-sidebarItem';

// https://github.com/facebook/react/issues/7386#issuecomment-238091398
jest.mock('react-dom');
jest.mock('react-router-dom', () => ({
  Link: require('./mockComponent')('Link'),
}));

describe('SidebarItem', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <SidebarItem
          id="one"
          label="an/extremely/amazingly/ultra/hyper/really/very/long/label"
          icon="folder-o"
          link="some/url"
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly when hovered', () => {
    const tree = renderer
      .create(
        <SidebarItem
          id="two"
          label="two"
          icon="folder-o"
          link="some/url"
          hovering="two"
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly when selected', () => {
    const tree = renderer
      .create(
        <SidebarItem
          id="three"
          label="three"
          icon="folder-o"
          link="some/url"
          fSelected
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly when selected & hovered', () => {
    const tree = renderer
      .create(
        <SidebarItem
          id="four"
          label="four"
          icon="folder-o"
          link="some/url"
          hovering="four"
          fSelected
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('SidebarGroup', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <SidebarGroup name="Snapshot group name">{['one', 'two']}</SidebarGroup>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
