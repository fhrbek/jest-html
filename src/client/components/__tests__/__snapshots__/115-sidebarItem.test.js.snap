// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SidebarGroup renders correctly 1`] = `
<div
  style={
    Object {
      "backgroundColor": "white",
      "margin": "0.5em 0",
      "padding": "0.3em 0",
    }
  }
>
  <div
    style={
      Object {
        "color": "#999",
        "fontSize": "0.8em",
        "padding": "0.3em 1em",
        "textTransform": "uppercase",
      }
    }
  >
    Snapshot group name
  </div>
  one
  two
</div>
------------HTML PREVIEW---------------
<div style="background-color:white;margin:0.5em 0;padding:0.3em 0;">
  <div style="color:#999;font-size:0.8em;padding:0.3em 1em;text-transform:uppercase;">
    Snapshot group name
  </div>
  one
  two
</div>
`;

exports[`SidebarItem renders correctly 1`] = `
<div
  dataMockType="Link"
  id="one"
  style={
    Object {
      "alignItems": "center",
      "color": "currentColor",
      "cursor": "pointer",
      "display": "flex",
      "flexDirection": "row",
      "padding": "0.3em 1em",
      "textDecoration": "none",
      "wordBreak": "break-word",
    }
  }
  to="some/url"
>
  <div
    style={
      Object {
        "color": "blue",
        "fontWeight": "bold",
      }
    }
  >
    Link
  </div>
  <div
    style={
      Object {
        "WebkitFlex": "0 0 20px",
        "flex": "0 0 20px",
      }
    }
  >
    <i
      className="giu-icon fa fa-folder-o"
      style={
        Object {
          "color": undefined,
          "cursor": "pointer",
          "fontSize": undefined,
          "letterSpacing": "normal",
        }
      }
    />
  </div>
  <div
    style={
      Object {
        "WebkitFlex": 1,
        "flex": 1,
      }
    }
  >
    an/extremely/amazingly/ultra/hyper/really/very/long/label
     
  </div>
</div>
------------HTML PREVIEW---------------
<div
  dataMockType="Link"
  id="one"
  style="align-items:center;color:currentColor;cursor:pointer;display:flex;flex-direction:row;padding:0.3em 1em;text-decoration:none;word-break:break-word;"
  to="some/url"
>
  <div style="color:blue;font-weight:bold;">
    Link
  </div>
  <div style="-webkit-flex:0 0 20px;flex:0 0 20px;">
    <i
      class="giu-icon fa fa-folder-o"
      style="cursor:pointer;letter-spacing:normal;"
    >
    </i>
  </div>
  <div style="-webkit-flex:1;flex:1;">
    an/extremely/amazingly/ultra/hyper/really/very/long/label
     
  </div>
</div>
`;

exports[`SidebarItem renders correctly when hovered 1`] = `
<div
  dataMockType="Link"
  id="two"
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#e6e6e6",
      "color": "currentColor",
      "cursor": "pointer",
      "display": "flex",
      "flexDirection": "row",
      "padding": "0.3em 1em",
      "textDecoration": "none",
      "wordBreak": "break-word",
    }
  }
  to="some/url"
>
  <div
    style={
      Object {
        "color": "blue",
        "fontWeight": "bold",
      }
    }
  >
    Link
  </div>
  <div
    style={
      Object {
        "WebkitFlex": "0 0 20px",
        "flex": "0 0 20px",
      }
    }
  >
    <i
      className="giu-icon fa fa-folder-o"
      style={
        Object {
          "color": undefined,
          "cursor": "pointer",
          "fontSize": undefined,
          "letterSpacing": "normal",
        }
      }
    />
  </div>
  <div
    style={
      Object {
        "WebkitFlex": 1,
        "flex": 1,
      }
    }
  >
    two
     
  </div>
</div>
------------HTML PREVIEW---------------
<div
  dataMockType="Link"
  id="two"
  style="align-items:center;background-color:#e6e6e6;color:currentColor;cursor:pointer;display:flex;flex-direction:row;padding:0.3em 1em;text-decoration:none;word-break:break-word;"
  to="some/url"
>
  <div style="color:blue;font-weight:bold;">
    Link
  </div>
  <div style="-webkit-flex:0 0 20px;flex:0 0 20px;">
    <i
      class="giu-icon fa fa-folder-o"
      style="cursor:pointer;letter-spacing:normal;"
    >
    </i>
  </div>
  <div style="-webkit-flex:1;flex:1;">
    two
     
  </div>
</div>
`;

exports[`SidebarItem renders correctly when selected & hovered 1`] = `
<div
  dataMockType="Link"
  id="four"
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#9d3614",
      "color": "white",
      "cursor": "pointer",
      "display": "flex",
      "flexDirection": "row",
      "padding": "0.3em 1em",
      "textDecoration": "none",
      "wordBreak": "break-word",
    }
  }
  to="some/url"
>
  <div
    style={
      Object {
        "color": "blue",
        "fontWeight": "bold",
      }
    }
  >
    Link
  </div>
  <div
    style={
      Object {
        "WebkitFlex": "0 0 20px",
        "flex": "0 0 20px",
      }
    }
  >
    <i
      className="giu-icon fa fa-folder-o"
      style={
        Object {
          "color": undefined,
          "cursor": "pointer",
          "fontSize": undefined,
          "letterSpacing": "normal",
        }
      }
    />
  </div>
  <div
    style={
      Object {
        "WebkitFlex": 1,
        "flex": 1,
      }
    }
  >
    four
     
  </div>
</div>
------------HTML PREVIEW---------------
<div
  dataMockType="Link"
  id="four"
  style="align-items:center;background-color:#9d3614;color:white;cursor:pointer;display:flex;flex-direction:row;padding:0.3em 1em;text-decoration:none;word-break:break-word;"
  to="some/url"
>
  <div style="color:blue;font-weight:bold;">
    Link
  </div>
  <div style="-webkit-flex:0 0 20px;flex:0 0 20px;">
    <i
      class="giu-icon fa fa-folder-o"
      style="cursor:pointer;letter-spacing:normal;"
    >
    </i>
  </div>
  <div style="-webkit-flex:1;flex:1;">
    four
     
  </div>
</div>
`;

exports[`SidebarItem renders correctly when selected 1`] = `
<div
  dataMockType="Link"
  id="three"
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#ca461a",
      "color": "white",
      "cursor": "pointer",
      "display": "flex",
      "flexDirection": "row",
      "padding": "0.3em 1em",
      "textDecoration": "none",
      "wordBreak": "break-word",
    }
  }
  to="some/url"
>
  <div
    style={
      Object {
        "color": "blue",
        "fontWeight": "bold",
      }
    }
  >
    Link
  </div>
  <div
    style={
      Object {
        "WebkitFlex": "0 0 20px",
        "flex": "0 0 20px",
      }
    }
  >
    <i
      className="giu-icon fa fa-folder-o"
      style={
        Object {
          "color": undefined,
          "cursor": "pointer",
          "fontSize": undefined,
          "letterSpacing": "normal",
        }
      }
    />
  </div>
  <div
    style={
      Object {
        "WebkitFlex": 1,
        "flex": 1,
      }
    }
  >
    three
     
  </div>
</div>
------------HTML PREVIEW---------------
<div
  dataMockType="Link"
  id="three"
  style="align-items:center;background-color:#ca461a;color:white;cursor:pointer;display:flex;flex-direction:row;padding:0.3em 1em;text-decoration:none;word-break:break-word;"
  to="some/url"
>
  <div style="color:blue;font-weight:bold;">
    Link
  </div>
  <div style="-webkit-flex:0 0 20px;flex:0 0 20px;">
    <i
      class="giu-icon fa fa-folder-o"
      style="cursor:pointer;letter-spacing:normal;"
    >
    </i>
  </div>
  <div style="-webkit-flex:1;flex:1;">
    three
     
  </div>
</div>
`;
